pkg/trace/traceutil/otel_util.go

```diff
@@ -0,0 +1,277 @@
+	semconv117 "go.opentelemetry.io/collector/semconv/v1.17.0"
+	semconv "go.opentelemetry.io/collector/semconv/v1.5.0"
+	"go.opentelemetry.io/otel/attribute"
+)
+
+// Util functions for converting OTel semantics to DD semantics.
+// TODO(OTEL-1726): reuse the same mapping code for ReceiveResourceSpans and Concentrator
+
+var (
+	// SignalTypeSet is the OTel attribute set for traces.
+	SignalTypeSet = attribute.NewSet(attribute.String("signal", "traces"))
+)
+
+// IndexOTelSpans iterates over the input OTel spans and returns 3 maps:
+// OTel spans indexed by span ID, OTel resources indexed by span ID, OTel instrumentation scopes indexed by span ID.
+// Skips spans with invalid trace ID or span ID. If there are multiple spans with the same (non-zero) span ID, the last one wins.
+func IndexOTelSpans(traces ptrace.Traces) (map[pcommon.SpanID]ptrace.Span, map[pcommon.SpanID]pcommon.Resource, map[pcommon.SpanID]pcommon.InstrumentationScope) {
+	spanByID := make(map[pcommon.SpanID]ptrace.Span)
+	resByID := make(map[pcommon.SpanID]pcommon.Resource)
+	scopeByID := make(map[pcommon.SpanID]pcommon.InstrumentationScope)
+	rspanss := traces.ResourceSpans()
+	for i := 0; i < rspanss.Len(); i++ {
+		rspans := rspanss.At(i)
+		res := rspans.Resource()
+		for j := 0; j < rspans.ScopeSpans().Len(); j++ {
+			libspans := rspans.ScopeSpans().At(j)
+			for k := 0; k < libspans.Spans().Len(); k++ {
+				span := libspans.Spans().At(k)
+				if span.TraceID().IsEmpty() || span.SpanID().IsEmpty() {
+					continue
+				}
+				spanByID[span.SpanID()] = span
+				resByID[span.SpanID()] = res
+				scopeByID[span.SpanID()] = libspans.Scope()
+			}
+		}
+	}
+	return spanByID, resByID, scopeByID
+}
+
+// GetTopLevelOTelSpans returns the span IDs of the top level OTel spans.
+func GetTopLevelOTelSpans(spanByID map[pcommon.SpanID]ptrace.Span, resByID map[pcommon.SpanID]pcommon.Resource, topLevelByKind bool) map[pcommon.SpanID]struct{} {
+	topLevelSpans := make(map[pcommon.SpanID]struct{})
+	for spanID, span := range spanByID {
+		if span.ParentSpanID().IsEmpty() {
+			// case 1: root span
+			topLevelSpans[spanID] = struct{}{}
+		} else if topLevelByKind {
+			// New behavior for computing top level OTel spans, see computeTopLevelAndMeasured in pkg/trace/api/otlp.go
+			spanKind := span.Kind()
+			if spanKind == ptrace.SpanKindServer || spanKind == ptrace.SpanKindConsumer {
+				// span is a server-side span, mark as top level
+				topLevelSpans[spanID] = struct{}{}
+			}
+		} else {
+			// Fall back to old behavior in ComputeTopLevel
+			if _, ok := spanByID[span.ParentSpanID()]; !ok {
+				// case 2: parent span not in the same chunk, presumably it belongs to another service
+				topLevelSpans[spanID] = struct{}{}
+			} else {
+				svc := GetOTelAttrVal(resByID[spanID].Attributes(), true, semconv.AttributeServiceName)
+				parentSvc := GetOTelAttrVal(resByID[span.ParentSpanID()].Attributes(), true, semconv.AttributeServiceName)
+				if svc != parentSvc {
```

Comment:
nit: Would prefer to reduce the amount of nested `if` statements in this function, maybe this could be done by using `continue` at the end of some `if` statements?

Reply:
➕

Reply:
Sounds reasonable, done